### ISSUE TITLE
Wake On Lan

### ISSUE THREAD

**Issue by Timmguin** [AlexGustafsson/homebridge-wol#129]
Hello so i am completely new to github and anything involving coding but i do have wake on lan set up with my pc and alexa via an alexa skill but i will be switching to apple home kit soon and i would love to get this feature working there as well and i found this and i think It's is what i need and i added the plugin to my homebridge but i'm having trouble understanding how to set it up so help would be very much appreciated :)

**Comment by AlexGustafsson** [comment]
I'll try to cover each step, but you're most likely familiar with some parts already so just skip those you've already taken care of.

I'm assuming you've got homebridge installed. If not, see https://github.com/homebridge/homebridge/wiki.

The first step to get Wake On LAN up and running is to identify your target device's MAC address. The following resource should cover that: https://www.wikihow.com/Find-the-MAC-Address-of-Your-Computer.

Next, homebridge will need to be configured. If you haven't installed the plugin yet you can do so either via the Homebridge UI if installed, or via `npm install -g homebridge-wol`.  To configure a device for Wake On LAN, simply add the following to your homebridge installation's `config.json` in the `accessories` array:

```json
{
  "accessory": "NetworkDevice",
  "name": "Generic",
  "mac": "xx:xx:xx:xx:xx:xx"
}
```

Here `accessory` is required to be `NetworkDevice`, `name` can be set to anything and `mac` should be set to the MAC address you found from the first step. 

This will make your PC show up as a switch in HomeKit and you will be able to turn it on. Next, you might want the switch to actually reflect the state of your PC - whether it's turned on or off. You'll then need to find its IP address (https://www.whatismybrowser.com/detect/what-is-my-local-ip-address). 

Update your accessory configuration to add your IP in the `ip` field:

```json
{
  "accessory": "NetworkDevice",
  "name": "Generic",
  "mac": "xx:xx:xx:xx:xx:xx",
  "ip": "192.168.1.xxx"
}
```

The switch will now reflect the state of your PC.

Lastly, there is a lot of tweaking and other features available, but none of which are necessary. These are documented in the `README` and also available in the Homebridge UI.

**Comment by Timmguin** [comment]
Hello so i did all the things listed above but my computer doesn't turn on or off it shows up in home kit but doesn't do anything when i tap it i think this is because i did not enter anything in the homebridge field that says _**Start command**_ and **_Shutdown command_** but i dont know what to enter.

I have included the logs below dont know if it helps 
I replaced my ip and time and date with x


[xx/xx/xxxx, xx:xx:xx] [PC] NetworkDevice shutdown cycle started for "PC" (xxx.xxx.x.xxx)
[xx/xx/xxxx, xx:xx:xx] [PC] NetworkDevice "PC" (xxx.xxx.x.xxx) went from status "Online" to "Shutting Down"
[xx/xx/xxxx, xx:xx:xx] [PC] No way of shutting down "PC" (xxx.xxx.x.xxx) was configured
[xx/xx/xxxx, xx:xx:xx] [PC] NetworkDevice "PC" (xxx.xxx.x.xxx) went from status "Shutting Down" to "Online"

**Comment by AlexGustafsson** [comment]
Starting a computer using wake on lan only requires the MAC. The start and shutdown commands are for more complex uses where you’d want more custom behaviour. You will also make sure that your computer is set up to actually work with WoL: https://www.lifewire.com/wake-on-lan-4149800

If starting still does not work after ensuring that WoL is enabled, then there might be something acting up in your network. If that’s the case, we’ll try to do more investigative work.

As for the shutdown command (necessary to actually shut down a computer), try to read the community wiki guide for some clues: https://github.com/AlexGustafsson/homebridge-wol/wiki/Running-on-Synology-Docker-and-turn-Windows-10-on-off-using-SSH-and-Samba

As a general note, the various start and shutdown commands are simply any command that can be run on the system in a regular terminal such as CMD in Windows. This plugin doesn’t intervene with the command in any way. For help with those, it’s usually preferable to search for it on a stack exchange site such as super user.

**Comment by Timmguin** [comment]
I have now managed to get it to turn on using home kit but turning it off doesn't seem to work yet and im not sure why.
Turning it off using the alexa skill works just fine.

**Comment by AlexGustafsson** [comment]
@Timmguin can you share the `shutdownCommand` you're using? Make sure you strip any personal information. Also, could you share the logs? The samba tool used to communicate with Windows PC might have to be installed on your server running homebridge. The logs will yell if there's any issue executing the shutdown command.

**Comment by Timmguin** [comment]
I did some reserch myself and i found out that /s shuts down the mashine but it dosent work.


I have included the logs below
I replaced my ip and time/date with x

    [xx/xx/xxxx, xx:xx:xx] [PC] NetworkDevice shutdown cycle started for "PC" (xxx.xxx.x.xxx)
    [xx/xx/xxxx, xx:xx:xx] [PC] NetworkDevice "PC" (xxx.xxx.x.xxx) went from status "Online" to "Shutting Down"
    [xx/xx/xxxx, xx:xx:xx] [PC] Attempting to shut down "PC" (xxx.xxx.x.xxx) using "shutdown"
    [xx/xx/xxxx, xx:xx:xx] [PC] An error occured while trying to shut down "PC" (xxx.xxx.x.xxx): Error: Command failed: shutdown /s
    Failed to parse time specification: /s

        at ChildProcess.exithandler (child_process.js:308:12)
        at ChildProcess.emit (events.js:315:20)
        at maybeClose (internal/child_process.js:1048:16)
        at Process.ChildProcess._handle.onexit (internal/child_process.js:288:5) {
      killed: false,
      code: 1,
      signal: null,
      cmd: 'shutdown /s',
      stdout: '',
      stderr: 'Failed to parse time specification: /s\n'
    }
    [xx/xx/xxxx, xx:xx:xx] [PC] Waiting for shutdown grace time (15000ms) for "PC" (xxx.xxx.x.xxx)
    [xx/xx/xxxx, xx:xx:xx] [PC] NetworkDevice "PC" (xxx.xxx.x.xxx) went from status "Shutting Down" to "Online"

**Comment by AlexGustafsson** [comment]
As you can see, there are a lot of logs after the shutdown command is ran. You’ll have to debug the shutdown command, since it seems to be incorrect. It’s always a good idea to run the command outside of this plugin. The plugin does nothing with the command, so if it works in a regular terminal, it will work with the plugin.

**Comment by Timmguin** [comment]
@AlexGustafsson Since i dont have any experience with the terminal or coding do you have any idea on what shutdown command i could try ?

**Comment by AlexGustafsson** [comment]
Try to follow this guide: https://github.com/AlexGustafsson/homebridge-wol/wiki/Running-on-Synology-Docker-and-turn-Windows-10-on-off-using-SSH-and-Samba.

**Comment by Timmguin** [comment]
Hi so did my absolute best trying to follow the tutorial but at some point i got lost i know this is a lot to ask for but maybe if you have discord or something similar maybe we could hop on a call or just chat about it privately :)

**Comment by AlexGustafsson** [comment]
Sorry, but I don't have time for that. I'll leave this issue open, perhaps someone have something to chip in. You can always try to ask on https://superuser.com, Reddit or the like since the issue is quite general and not directly related to this plugin, there should be people there able to help you in a better way.

If you're interested in the topics of a terminal etc. you can always explore it more on your own and come back to the guide to experiment if you've learnt more.

Unfortunately I don't think there's any real way for this plugin to improve the usability as there will always be manual setup required for various computers and operating system, something that's out of scope for this simple plugin. If you have any ideas, please let me know.

Good luck and feel free to post updates here if you are able to progress or solve the issue - it might help others in the future!

**Comment by Timmguin** [comment]
Hello @AlexGustafsson 
so i have some updates regarding my issue so i managed to connect via SSH but it's still not quite working yet but i dont understand whats going wrong i have attached the logs i really hope you can help me i think im pretty close to getting it to work.

    [xx/xx/xxxx, xx:xx:xx] [PC] NetworkDevice shutdown cycle started for "PC" (xxx.xxx.x.xxx)
    [xx/xx/xxxx, xx:xx:xx] [PC] NetworkDevice "PC" (xxx.xxx.x.xxx) went from status "Online" to "Shutting Down"
    [xx/xx/xxxx, xx:xx:xx] [PC] Attempting to shut down "PC" (xxx.xxx.x.xxx) using "ssh"
    [xx/xx/xxxx, xx:xx:xx] [PC] An error occured while trying to shut down "PC" (xxx.xxx.x.xxx): Error: Command failed: ssh -o 
    StrictHostKeyChecking=no -i /homebridge/id_rsa username@xxx.xxx.x.xxx shutdown /s
    Warning: Identity file /homebridge/id_rsa not accessible: No such file or directory.
    Permission denied, please try again.
    Permission denied, please try again.
    username@xxx.xxx.x.xxx: Permission denied (publickey,password,keyboard-interactive).
    
        at ChildProcess.exithandler (child_process.js:308:12)
        at ChildProcess.emit (events.js:315:20)
        at maybeClose (internal/child_process.js:1048:16)
        at Process.ChildProcess._handle.onexit (internal/child_process.js:288:5) {
      killed: false,
      code: 255,
      signal: null,
      cmd: 'ssh -o StrictHostKeyChecking=no -i /homebridge/id_rsa timm@192.168.1.123 shutdown /s',
      stdout: '',
      stderr: 'Warning: Identity file /homebridge/id_rsa not accessible: No such file or directory.\n' +
        'Permission denied, please try again.\r\n' +
        'Permission denied, please try again.\r\n' +
        'username@xxx.xxx.x.xxx: Permission denied (publickey,password,keyboard-interactive).\r\n'
    }
    [xx/xx/xxxx, xx:xx:xx] [PC] Waiting for shutdown grace time (15000ms) for "PC" (xxx.xxx.x.xxx)
    [xx/xx/xxxx, xx:xx:xx] [PC] NetworkDevice "PC" (xxx.xxx.x.xxx) went from status "Shutting Down" to "Online"

**Comment by AlexGustafsson** [comment]
It looks like the RSA key has not been created - or it was created under another user. Make sure it exists or try to create it (there should be steps for this in the aforementioned guide).

From the logs:
```
Warning: Identity file /homebridge/id_rsa not accessible: No such file or directory.
```

**Comment by Timmguin** [comment]
i looked at the guide and spoke to some people on discord but they didn't know what was causing the problem any idea on what i could try ?

**Comment by Timmguin** [comment]
**Update**

So i now successfully shut down my pc using homebridge for the first time but only by using the terminal and entering the password when i click the button on homebridge for homekit nothing happens these are the logs.

I think it has something to do with the password since i had to enter it manually in the homebridge terminal in order for it to work.


    [xx/xx/xxxx, xx:xx:xx] [PC] NetworkDevice shutdown cycle started for "PC" (xxx.xxx.x.xxx)
    [xx/xx/xxxx, xx:xx:xx] [PC] NetworkDevice "PC" (xxx.xxx.x.xxx) went from status "Online" to "Shutting Down"
    [xx/xx/xxxx, xx:xx:xx] [PC] Attempting to shut down "PC" (xxx.xxx.x.xxx) using "ssh"
    [xx/xx/xxxx, xx:xx:xx] [PC] An error occured while trying to shut down "PC" (xxx.xxx.x.xxx): Error: Command failed: ssh -o 
    StrictHostKeyChecking=no -i /homebridge/id_rsa username@xxx.xxx.x.xxx shutdown /s
    Warning: Identity file /homebridge/id_rsa not accessible: No such file or directory.
    Permission denied, please try again.
    Permission denied, please try again.
    username@xxx.xxx.x.xxx: Permission denied (publickey,password,keyboard-interactive).
    
        at ChildProcess.exithandler (child_process.js:308:12)
        at ChildProcess.emit (events.js:315:20)
        at maybeClose (internal/child_process.js:1048:16)
        at Socket.<anonymous> (internal/child_process.js:439:11)
        at Socket.emit (events.js:315:20)
        at Pipe.<anonymous> (net.js:673:12) {
      killed: false,
      code: 255,
      signal: null,
      cmd: 'ssh -o StrictHostKeyChecking=no -i /homebridge/id_rsa username@xxx.xxx.x.xxx shutdown /s',
      stdout: '',
      stderr: 'Warning: Identity file /homebridge/id_rsa not accessible: No such file or directory.\n' +
        'Permission denied, please try again.\r\n' +
        'Permission denied, please try again.\r\n' +
        'username@xxx.xxx.x.xxx: Permission denied (publickey,password,keyboard-interactive).\r\n'
    }
    [xx/xx/xxxx, xx:xx:xx] [PC] Waiting for shutdown grace time (15000ms) for "PC" (xxx.xxx.x.xxx)
    [xx/xx/xxxx, xx:xx:xx] [PC] NetworkDevice "PC" (xxx.xxx.x.xxx) went from status "Shutting Down" to "Online"

Thanks for all te help i have gotten so far i think im close to getting it to work but at this point i dont know what to try so help would be very much appreciated :)

**Comment by AlexGustafsson** [comment]
You look to be close now. The last step is to ensure that you're authenticating using SSH keys.

As the user running Homebridge on your server, try to run `ssh -i /homebridge/id_rsa username@remote_Address echo 'hello world!'` to connect to your Windows machine and print "hello world!". If you're required to enter your password, SSH keys are not setup properly and you should continue to read this.

First, check if you have configured SSH keys. Do this by running the following command as the user running homebridge:

```
ls /homebridge/id_rsa
```

If you receive "no such file or directory", you will have to set the keys up.

```
> mkdir /homebridge
> ssh-keygen
Generating public/private rsa key pair.
Enter file in which to save the key (/root/.ssh/id_rsa): /homebridge/id_rsa
Enter passphrase (empty for no passphrase):
Enter same passphrase again:
Your identification has been saved in /homebridge/id_rsa
Your public key has been saved in /homebridge/id_rsa.pub
```

Basically, try to look through the guide for the steps covering SSH keys.

**Comment by Timmguin** [comment]
So i got to the point where i have to enter the passphrase and i left both of them empty but then i get this.

    Saving key "/homebridge/id_rsa" failed: Permission denied

**Comment by AlexGustafsson** [comment]
You do not seem to have the privilege to create that file. You may either chose to create the file as root (with `sudo` prefixed to the command), but unless you're already running homebridge as root, you're better off creating the file in a directory the user has the privilege to use. For example, if your homebridge user is `homebridge`, you might as well exchange `/homebridge/id_rsa` with `/home/homebridge/.ssh/homebridge_rsa` etc.

**Comment by Timmguin** [comment]
@AlexGustafsson I finally got it to work so now i generated the SSH keys and saved them under /homebridge/id_rsa what is going to be the next step ?

I looked through the guide but i dont know how i can get the file from the raspberry pi on to my pc because i dont have a nas or know what it is.

**Comment by Timmguin** [comment]
@AlexGustafsson So i got the file on to my pc with a usb drive and did all the things in the tutorial but it still asks for a password when i try to connect.

**Comment by AlexGustafsson** [comment]
On the homebridge server you should now have `/homebridge/id_rsa` and `/homebridge/id_rsa.pub`. To setup the target computer, run the following command on the homebridge server:

```
ssh-copy-id -i /homebridge/id_rsa user@host
```

Here, `user` is the username you use to login to your Windows PC and `host` the IP address of the Windows PC. You should be prompted to use your password. The command will likely write a text like "Succeeded, now try to login without using a password".

To try to login without using a password, you can now simply execute:

```
ssh -i /homebridge/id_rsa user@host
```

You do not need to do anything manually on your Windows PC. So for safety you should be able to remove any files you've copied over. In this case, you can think of `/homebridge/id_rsa` as your password and `/homebridge/id_rsa.pub` to be your username. The `ssh-copy-id` command will copy over `/homebridge/id_rsa.pub` to the target and set it up to allow you to login using the key instead of a password.

If you are able to login using the command above without any password, your homebridge setup should work as well.

**Comment by Timmguin** [comment]
When i type in the first command i get this.


    pi@homebridgedhomebridge $ ssh-copy-id /homebridge/id_rsa username@xxx.xxx.x.xxx
    /usr/bin/ssh-copy-id: ERROR: failed to open ID file '/homebridge/id rsa': Permission denied (to install the contents of 
    '/homebridge/id_rsa.pub' anyway, look at the -f option)


I replaced my username and IP with (username@xxx.xxx.x.xxx)

**Comment by AlexGustafsson** [comment]
Try using `sudo`: `sudo ssh-copy-id /homebridge/id_rsa username@xxx.xxx.x.xxx`

**Comment by Timmguin** [comment]
Im still getting the same thing.

**Comment by AlexGustafsson** [comment]
Did you try the suggested `-f` option?

```
ssh-copy-id -f /homebridge/id_rsa username@xxx.xxx.x.xxx
```

**Comment by Timmguin** [comment]
I'm still getting this.

    /usr/bin/ssh-copy-id: ERROR: Too many arguments. Expecting a target hostname, got: 
    1/homebridge/id_rsa"username@xxx.xxx.x.xxx' 
    Usage: /usr/bin/ssh-copy-id [-h1-21-fl-n] [-i [identity_file]] [-p port] [(-o <ssh -o options>] ...] [userflhostname -f: force 
    mode -- copy keys without trying to check if they are already installed -n: dry run -- no keys are actually copied -hi-?: print 
    this help

**Comment by Timmguin** [comment]
@AlexGustafsson Is there anything else i can try ?

**Comment by AlexGustafsson** [comment]
I'd recommend reading up on the following topics and potentially post questions on https://superuser.com or https://askubuntu.com if there are things you may not understand. There are also some keywords you may use to find more information online.

1. SSH key authentication - how to set up ssh keys
2. Setting up OpenSSH on Windows 10
3. SSH key authentication on Windows 10
4. SSH specifying SSH identity to use

Unfortunately, it's a non-goal of this project to serve as an end-to-end solution for controlling computers via HomeKit as that would require further applications etc. installed on each device. Right now our solution is very general, allowing any command, but as you've experienced first hand it may be quite the hurdle to overcome depending on your previous knowledge. Hopefully, the documentation for common use cases like yours will become better with time, but right now I don't have the time to write it on my own. Thankfully I have received help with issues and wikis in the past and hopefully it will continue.

If you're able to solve your issue, please come back and write about it here or as a wiki post as it can help future users facing the same issues. If you have further questions more specific to this plugin or its usage, don't hesitate to ask them here as a discussion post (https://github.com/AlexGustafsson/homebridge-wol/discussions). If you have any feature requests or if you find bugs, you're always welcome top open an issue. Good luck!
